eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/BitcoinCoreWallet.scala

```diff
@@ -62,15 +61,15 @@ class BitcoinCoreWallet(rpcClient: BitcoinJsonRPCClient)(implicit ec: ExecutionC
 
   def signTransaction(tx: Transaction): Future[SignTransactionResponse] = signTransaction(Transaction.write(tx).toHex)
 
-  def getTransaction(txid: ByteVector32): Future[Transaction] = rpcClient.invoke("getrawtransaction", txid.toString()) collect { case JString(hex) => Transaction.read(hex) }
+  def getTransaction(txid: ByteVector32): Future[Transaction] = bitcoinClient.getTransaction(txid)
 
-  def publishTransaction(tx: Transaction)(implicit ec: ExecutionContext): Future[String] = publishTransaction(Transaction.write(tx).toHex)
+  def publishTransaction(tx: Transaction)(implicit ec: ExecutionContext): Future[String] = bitcoinClient.publishTransaction(tx)
 
-  def publishTransaction(hex: String)(implicit ec: ExecutionContext): Future[String] = rpcClient.invoke("sendrawtransaction", hex) collect { case JString(txid) => txid }
+  def publishTransaction(hex: String)(implicit ec: ExecutionContext): Future[String] = publishTransaction(Transaction.read(hex))
 
-  def unlockOutpoints(outPoints: Seq[OutPoint])(implicit ec: ExecutionContext): Future[Boolean] = rpcClient.invoke("lockunspent", true, outPoints.toList.map(outPoint => Utxo(outPoint.txid.toString, outPoint.index))) collect { case JBool(result) => result }
+  def unlockOutpoints(outPoints: Seq[OutPoint])(implicit ec: ExecutionContext): Future[Boolean] = rpcClient.invoke("lockunspent", true, outPoints.toList.map(outPoint => Utxo(outPoint.txid, outPoint.index))) collect { case JBool(result) => result }
```

Comment:
It's a bit hard to understand how functionality is divided between `BitcoinCoreWallet` and `ExtendedBitcoinClient`.

The way I understand it, `ExtendedBitcoinClient` is just a thin typed wrapper around bitcoin-cli RPC, is that correct (with the exception of the kinda weird channel announcement validation)?
And `BitcoinCoreWallet` adds some lightning-specific functionality such as composing multiple calls to the `ExtendedBitcoinClient` to do things like creating the funding tx.

If that's indeed how it was meant to be split, it feels to me that `unlockOutpoints`, `getBalance` and `getNewAddress` should be provided by `ExtendedBitcoinClient`. I'm wondering if we shouldn't take this opportunity to provide cleaner interfaces, but I don't know that part of the codebase well so maybe it's not worth it?

Reply:
That's correct, `ExtendedBitcoinClient` is a thin typed wrapper around `BitcoinJsonRPCClient` that exposes common functionalities used by other parts of the codebase, for example to get the tx confirmations we don't directly use the RPC client (and process the response) but we'd  reuse the code in `ExtendedBitcoinClient`. `BitcoinCoreWallet` is an implementation of `EclairWallet` interface with the bitcon-core backend, it naturally uses the `ExtendedBitcoinClient` for most operations ([example](https://github.com/ACINQ/eclair/pull/1248/files#diff-f0e4b1153995aaf8ef920c12f6c4c36bR64)) but it also contains lightning-specific logic for making the funding transaction etc. Having considered this, the `unlockOutpoints` seems a good fit for `BitcoinCoreWallet` because it's used only there and is specific to the bitcoin-core behavior, WDYT?